### PR TITLE
Implemented FM and created a test for it

### LINKED ISSUE #97: Converting gromacs all-atom trajectory to a hooomd gsd file
Need a thorough documentation and a python script to convert an all-atom trajectory to a hoomd trajectory file.

### LINKED ISSUE #106: Implement force matching as method

### PR DESCRIPTION
**Pull request by hgandhi2411**
resolves #106

### CODE REVIEW

**Comment by whitead** [comment]
@hgandhi2411 Can you tell us about the goal of this code? Is it to run htf on a trajectory, so that you can compute quantities, or is it to convert a trajectory into a gsd file?

**Comment by hgandhi2411** [comment]
> @hgandhi2411 Can you tell us about the goal of this code? Is it to run htf on a trajectory, so that you can compute quantities, or is it to convert a trajectory into a gsd file?

The goal is to convert the trajectory into a gsd file.

**Comment by whitead** [comment]
@hgandhi2411 I think it might make more sense to actually just run hoomd-tf on the trajectory directly. Like we can't do anything with a gsd file, so what do we accomplish by converting it? The top code you have loads the trajectory into hoomd-tf and then you can do anything (e.g., compute CVs, match forces) so I don't see why the creation of the gsd file is necessary. Maybe @dilnoza92 has an opinion too on this.

**Comment by whitead** [comment]
I'm thinking kind of like this:

```python
# load trajectory
u
#
graph = htf.graph_builder(NN, use_trajectory=u)
...
...build_graph
out_nodes = [force_match_optimizer, energy]
...

graph.run_from_trajectory(nsteps, period, out_nodes=out_nodes)
```

File: htf/utils.py [hunk 1]
```diff
@@ -266,6 +266,22 @@ def sparse_mapping(molecule_mapping, molecule_mapping_index,
...
 
+def force_matching(cg_beads, mapped_forces, calculated_cg_forces):
+    M = cg_beads
+    # Make sure shape(mapped_forces) = shape(calculated_cg_forces) = [M x 3]
+    if not ((mapped_forces.shape.as_list() == [M, 3]) and
+            (calculated_cg_forces.shape.as_list() == [M, 3])):
+        raise ValueError('mapped_forces and calculated_cg_forces'\
+						 'must have dimension [cg_beads, 3]')
+    # minimize mean squared error
+    calculated_forces = tf.get_variable('calculated-forces', shape=[M, 3])
```
Comment (whitead) [inline htf/utils.py]: Why not use the forces directly? Is there a reason for the variable?

File: htf/models/example-models/traj-parse/traj_parser.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import hoomd.htf as htf
```
Comment (whitead) [inline htf/models/example-models/traj-parse/traj_parser.py]: Do you think we can integrate this as part of the library or must this be a separate script?

Comment (hgandhi2411) [reply htf/models/example-models/traj-parse/traj_parser.py]: I'm working on integrating this as part of the library. Will push changes soon.

File: htf/utils.py (hunk 1 above)
Comment (hgandhi2411) [reply htf/utils.py]: `tf.get_variable()` makes the variable trainable and when the optimizer is run, it's value is updated. Using the forces directly wouldn't update the values unless they are given as a trainable tensor.

**Comment by whitead** [comment]
@hgandhi2411 do you have a unit test or example I can use?

File: htf/utils.py (hunk 1 above)
Comment (whitead) [reply htf/utils.py]: Yes but the training won't propagate if you use `assign`. You should just use the forces passed directly and let the user decide how to create their variables

**Comment by janash** [comment]
Hi Heta, 

I looked over the code PR and I have a few comments.

I would suggest breaking up the `run_from_trajectory` function. It is doing two very different things depending on the input to the `function` argument.

I understand the desire to put these together, as they both use the same code in preparation in at the beginning of the function.  However, having a single function which processes the universe object violates the Single Responsibility Principle (of SOLID software design). I would suggest breaking it up into two or three functions, maybe something like the following

1. `process_universe` - function which processes a universe object and converts it to a HOOMD system.  (https://github.com/ur-whitelab/hoomd-tf/blob/e2b74317b062345252771f29aff76a04e8989af1/htf/utils.py#L276-L294). You'd have to think about what to return here, depending on what information you will need.
2. `write_gsd` - function which writes a gsd file from information output from `process_universe`. This may not be used ever by a user, it could be something internal, and you could have a third function which uses `process_universe` and `write_gsd` together to write a gsd file from a universe object. I'm suggesting breaking it up so that you can have another function use output from `process_universe` if you need to. 

For the force matching function, the way you do this may depend on some other things. Why does 'force_matching' called from `run_from_trajectory` take different arguments than the `force_matching` function? I'm sure this is something having to do with working with MDAnalysis universe object before the function is called, but it will need to be documented. 

Are you planning to implement more methods in addition to force matching?

**Comment by hgandhi2411** [comment]
Hi @janash,

Thanks for reviewing the PR and giving suggestions. I have separated the run_from_trajectory and force_matching functions. The process_trajectory part is only expected to be used by run_from_trajectory, so I will let that be as is.

File: htf/utils.py [hunk 3]
```diff
@@ -6,9 +6,10 @@
 from os import path
 import pickle
 import hoomd
+import gsd.hoomd
```
Comment (whitead) [inline htf/utils.py]: let's isolate this import to the one function it's needed for

File: htf/utils.py [hunk 4]
```diff
@@ -266,11 +266,132 @@ def sparse_mapping(molecule_mapping, molecule_mapping_index,
...
+def run_from_trajectory(model_directory, optimizer_tensor_name,
+                        loss_function_tensor_name, universe,
+                        selection='all', r_cut=10., period=10,
+                        feed_dict={}, out_nodes = []):
+    R""" This will process information from a trajectory and
+    run the user defined model on the nlist computed from the trajectory.
+
+    :param model_directory: The model directory
+    :type model_directory: string
+    :param optimizer_tensor_name: The tensor containing the optimizer
```
Comment (whitead) [inline htf/utils.py]: Rather than have an optimizer/loss, use the `out_nodes` variable in the model which indicates that tf ops to be run.

File: htf/utils.py [hunk 5]
```diff
@@ -266,11 +266,132 @@ def sparse_mapping(molecule_mapping, molecule_mapping_index,
...
+    N = (np.shape(type_array))[0]
+    NN = model_params['NN']
+    nlist = compute_nlist(atom_group.positions, r_cut=r_cut,
+                          NN=NN, system=system)
+    # Run the model at every nth frame, where n = period
+    with tf.Session() as sess:
+        for i, ts in enumerate(universe.trajectory):
+            positions = np.concatenate((atom_group.positions,
+                                        type_array), axis=1)
+            np_nlist = sess.run(nlist,
```
Comment (whitead) [inline htf/utils.py]: Should only need to have one step, not two

File: htf/utils.py [hunk 6]
```diff
@@ -266,11 +266,132 @@ def sparse_mapping(molecule_mapping, molecule_mapping_index,
...
+                                feed_dict={
+                                    graph.positions: positions,
+                                    graph.box: hoomd_box,
+                                    graph.batch_index: 0,
+                                    graph.batch_frac: 1})
+            optimizer, loss_function = sess.run([optimizer, loss_function],
+                                                feed_dict={
+                                                    **feed_dict,
+                                                    nlist_tensor: np_nlist})
+            graph.save(model_directory=model_directory,
```
Comment (whitead) [inline htf/utils.py]: Not sure why we would save the graph. Remember that the graph should already be saved. Maybe you meant to save a checkpoint? Try something like this in your session/for loop

```python
# at beginning
saver = tf.train.Saver()
# in loop
if i % period == 0:
  saver.save(sess, os.path.join(self.model_directory, 'model'),
                            global_step=i)
```

File: htf/utils.py [hunk 7]
```diff
@@ -266,11 +266,132 @@ def sparse_mapping(molecule_mapping, molecule_mapping_index,
...
+               move_previous=False)
+
+
+def force_matching(cg_beads, mapped_forces, calculated_cg_forces):
+    R""" This will minimize the difference between the mapped forces
+    and calculated CG forces using the Adam oprimizer and give updated
+    CG forces as a M x 3 tensor,
+
+    :param cg_beads: The number of coarse-grained (CG) beads in system
+    :type cg_beads: int
```
Comment (whitead) [inline htf/utils.py]: You should be able to infer this from `mapped_forces`

File: htf/utils.py [hunk 8]
```diff
@@ -266,11 +266,132 @@ def sparse_mapping(molecule_mapping, molecule_mapping_index,
...
+    # Make sure shape(mapped_forces) = shape(calculated_cg_forces) = [M x 3]
+    if not ((mapped_forces.shape.as_list() == [M, 3]) and
+            (calculated_cg_forces.shape.as_list() == [M, 3])):
+        raise ValueError('mapped_forces and calculated_cg_forces'
+                         'must have dimension [cg_beads, 3]')
+    # minimize mean squared error
+    cost = tf.losses.mean_squared_error(mapped_forces, calculated_cg_forces)
+    # It is assumed here that the user will pass in
+    # calculated_cg_forces that depend on trainable variables.
+    optimizer = tf.train.AdamOptimizer(learning_rate=1e-1).minimize(cost)
```
Comment (whitead) [inline htf/utils.py]: Make `learning_rate` an argument

File: htf/utils.py [hunk 9]
```diff
@@ -266,11 +266,132 @@ def sparse_mapping(molecule_mapping, molecule_mapping_index,
...
+    # load graph
+    graph = tf.train.import_meta_graph(path.join('{}/'.format(
+        model_directory), 'model.meta'), import_scope='')
+    with open('{}/graph_info.p'.format(model_directory), 'rb') as f:
+        model_params = pickle.load(f)
+    optimizer = tf.get_default_graph(
+    ).get_tensor_by_name(model_params[optimizer_tensor_name])
+    loss_function = tf.get_default_graph(
+    ).get_tensor_by_name(model_params[loss_function_tensor_name])
+    nlist_tensor = tf.get_default_graph(
```
Comment (whitead) [inline htf/utils.py]: Good

Comment (RainierBarrett) [reply htf/utils.py]: This code might read a bit easier if you put `tf.get_default_graph()` as a variable first

File: htf/utils.py [hunk 10]
```diff
@@ -266,11 +265,130 @@ def sparse_mapping(molecule_mapping, molecule_mapping_index,
...
+                               {'Lx': box[0],
+                                'Ly': box[1],
+                                'Lz': box[2]})})
+    hoomd_box = [[box[0], 0, 0], [0, box[1], 0], [0, 0, box[2]]]
+    # make type array
+    atom_group = universe.select_atoms(selection)
+    types = list(np.unique(atom_group.atoms.types))
+    type_array = np.array([types.index(i)
+                           for i in atom_group.atoms.types]).reshape(-1, 1)
+    N = (np.shape(type_array))[0]
```
Comment (whitead) [inline htf/utils.py]: This part is a bit confusing to me. Could you add a comment or two to explain what is happening?

File: htf/utils.py (hunk 5 above)
Comment (whitead) [reply htf/utils.py]: https://github.com/ur-whitelab/hoomd-tf/blob/master/htf/tfmanager.py#L424

File: htf/utils.py [hunk 11]
```diff
@@ -266,11 +265,135 @@ def sparse_mapping(molecule_mapping, molecule_mapping_index,
...
+    :type mapped_forces: tensor
+    :param calculated_cg_forces: A tensor with shape M x 3 where M is
+        number of CG beads in the system. These are CG forces estimated
+        using a function or a basis set.
+    :type calculated_cg_forces: tensor
+    :param learning_rate: The learning_rate for optimization
+    :type learning_rate: float
+
+    :return: The optimizer
+    :rtype: operation
```
Comment (RainierBarrett) [inline htf/utils.py]: [`AdamOptimizer` is of type `tf.compat.v1.train.Optimizer`](https://www.tensorflow.org/api_docs/python/tf/compat/v1/train/AdamOptimizer)

**Comment by whitead** [comment]
@hgandhi2411 is this ready for review again?

**Comment by whitead** [comment]
@hgandhi2411 can we remove seaborn as a dependency for the tutorial notebook?

**Comment by whitead** [comment]
@hgandhi2411 in the tutorial, please use `plt.figure()` each time you make a new figure. Also, the reason your trainable parameters aren't changing is that you have no forces in your simulation. Add an LJ interaction.

**Comment by hgandhi2411** [comment]
@whitead I made the changes. After adding LJ interaction, the cost fluctuations were smaller in magnitude but epsilon and sigma didn't change. So, I tried to bootstrap the variables but still, nothing changes.

**Comment by whitead** [comment]
@hgandhi2411 I have fixed the tutorial to force match correctly. I noticed you had a lot of unnecessary stuff going on like computing the true forces both in hoomd-blue and in htf and you were bootstrapping. Please see how I've set it up and make sure the documentation and unit tests follow this format. Also, could you reduce the default learning rate to 1e-3 or so. 1e-1 is too aggressive I think for default.

**Comment by whitead** [comment]
@hgandhi2411 also the optimizer/minimizer needs to be called in `out_nodes` to actually have the variables change.
